### PR TITLE
Support grain data checkpoint for elastic training

### PR DESCRIPTION
**Pull request by aireenmei**
# Description

* migrate RemoteIterator to colocated python class
* Add checkpointing logic to RemoteIterator, so data iterator in the colocated sidecar writes checkpoint to the checkpoint path, prevent sending data iterator state to the controller
* Add grain.ElasticIterator support controlled by flag grain_use_elastic_iterator, can be used with or without elastic training, with or without colocated_python. This class allows recovering checkpoint with a dynamic scale (up or down), with the following limitations, future work will loose the limitations:  
  (1). Only arrayrecord files are supported, parquet or tfrecord are not supported
  (2). Does not support many-to-one transformations, including packing, filtering
  (3). Does not support mixing datasets

# Tests
Tested on Pathways saving and restoring data iterator checkpoints with different # of v5e-32 slices
[jobset.yaml](https://paste.googleplex.com/5278394871840768)

### With colocated_python
`colocated_python_data_input=true colocated_python_checkpointing=true grain_use_elastic_iterator=true`, checkpoints in `gs://aireenmei-multipod/pathways-v5e/20260424_grain_elastic_2/aireen-pathways-v5e/checkpoints`:
* Start with 1 slice, set steps=25, checkpoint at step 0, 10, 20, 24: [log](https://cloudlogging.app.goo.gl/M5mH66AiYWMLNKRn9) (log confirms Num devices: 32), 
* Resume with 2 slices, set steps=45, checkpoint at step 30, 40, 44: [log](https://cloudlogging.app.goo.gl/Vcq6SufE1zqjnk3g8)(log confirms Num devices: 64), 
* Resume with 1 slice, set steps=65, checkpoint at step 50, 60, 64: [log](https://cloudlogging.app.goo.gl/56MKkZdQPchypPKT7)(log confirms Num devices: 32)

### Without colocated_python
`colocated_python_data_input=false colocated_python_checkpointing=false grain_use_elastic_iterator=true`, checkpoints in `gs://aireenmei-multipod/pathways-v5e/20260424_grain_elastic_3/aireen-pathways-v5e/checkpoints`:
* Start with 1 slice, set steps=25, checkpoint at step 0, 10, 20, 24: [log](https://cloudlogging.app.goo.gl/F7AP1yD26P3HoGdb6) (log confirms Num devices: 32), 
* Resume with 2 slices, set steps=45, checkpoint at step 30, 40, 44: [log](https://cloudlogging.app.goo.gl/Rk4QgfZ7tW6wdVUQ6)(log confirms Num devices: 64), 
* Resume with 1 slice, set steps=65, checkpoint at step 50, 60, 64: [log](https://cloudlogging.app.goo.gl/uHbfdxs4E31VuGmP7)(log confirms Num devices: 32)

### Index verification
Inspecting the checkpoints in `gs://aireenmei-multipod/pathways-v5e/20260424_grain_elastic_2/aireen-pathways-v5e/checkpoints/{step}/iter/process_0.json`
* Step 20: `"global_next_index": 672`
* Step 24: `"global_next_index": 800`
* Step 30: `"global_next_index": 1184`
* Step 20-24 is on 1 slice, 32 devices, batch_size=32, matches `(800-672) / 32 = 4 steps`
* Step 24-30 is on 2 slices, 64 devices, batch_size=64, matches `(1184-800) / 64 = 6 steps`

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Comment by lukebaumann** [comment]
How are you confirming that the dataset is resumed from the correct index?

File: src/maxtext/common/checkpointing.py [hunk 1]
```diff
@@ -94,6 +111,20 @@ def restore(
...
     process_count = getattr(args, "process_count", None)
 
+    # RemoteIteratorWrapper handles checkpointing via colocated python
+    if isinstance(item, RemoteIteratorWrapper):
+      step = int(directory.parent.name)
+      item.restore_state(step)
+      return item
+
+    # ElasticIterator: every process reads the same shared `process_0.json`.
+    if isinstance(item, ElasticIterator):
```
Comment (lukebaumann) [inline src/maxtext/common/checkpointing.py:122]: We want ElasitcIterator to also be a RemoteIterator. Is that happening?

File: src/maxtext/common/checkpointing.py [hunk 2]
```diff
@@ -790,22 +822,24 @@ def save_checkpoint(checkpoint_manager, step, state, config=None, data_iterator=
...
       data_iterator = [data_iterator]
-    grain_iters_to_save = []
-    process_count_total = jax.process_count() * len(data_iterator)
-    if config.expansion_factor_real_data > 1:
-      process_count_total = process_count_total // config.expansion_factor_real_data
-    for i, data_iter in enumerate(data_iterator):
-      process_index = jax.process_index() + i * jax.process_count()
-      grain_iters_to_save.append((data_iter.local_iterator, process_index, process_count_total))
-    save_args_composite["iter"] = GrainCheckpointSave(item=grain_iters_to_save)
+    if isinstance(data_iterator[0], RemoteIteratorWrapper):
```
Comment (SujeethJinesh) [inline src/maxtext/common/checkpointing.py]: Can you pull out what this actually is just for easier readability?

File: src/maxtext/input_pipeline/multihost_dataloading.py [hunk 3]
```diff
@@ -205,53 +172,120 @@ def _colocated_cpu_mesh(mesh: Mesh) -> Mesh:
...
 
-    return input_gdas
+  def __next__(self):
+    out = self.local_iterator.get_next(self.dummy_array)
+    # use tree_map is out is a dict
+    return jax.device_put(out, self.tpu_sharding)
+
+  def save_state(self, step):
+    step_array = jnp.full(self.dummy_array.shape, step, dtype=jnp.int32)
+    step_array = jax.device_put(step_array, self.cpu_sharding)
```
Comment (SujeethJinesh) [inline src/maxtext/input_pipeline/multihost_dataloading.py:285]: Do need any specialization?

**Comment by aireenmei** [comment]
> How are you confirming that the dataset is resumed from the correct index?

Good question. I just added a section "Index verification" in PR description.

File: src/maxtext/common/checkpointing.py (hunk 1 above)
Comment (aireenmei) [reply src/maxtext/common/checkpointing.py:122]: We want to support ElasitcIterator in both regular (Pathways, mcJAX) and colocated python environments. Because ElasitcIterator allows flexible chip counts, it has use case for regular customers who may change chip counts while keeping training progress. This specific path is for non-colocated cases (including Pathways and mcJAX), the Pathways + colocated case is handled by RemoteIteratorWrapper in the lines above. Let me improve the comments

File: src/maxtext/input_pipeline/multihost_dataloading.py (hunk 3 above)
Comment (aireenmei) [reply src/maxtext/input_pipeline/multihost_dataloading.py:285]: Could you explain specialization?

File: src/maxtext/common/checkpointing.py (hunk 2 above)
Comment (aireenmei) [reply src/maxtext/common/checkpointing.py]: I refactored and removed "[0]". Hopefully better readability now

File: src/maxtext/common/checkpointing.py (hunk 1 above)
Comment (lukebaumann) [reply src/maxtext/common/checkpointing.py:122]: RemoteIterator -> colocated python
RemoteIteratorWrapper -> elastic + colocated python
ElasticIterator -> elastic

All of these are usable for both McJAX and Pathways.

Is this correct?
Can the name RemoteIteratorWrapper be changed to ElasticRemoteIterator? Is there additional things the wrapper does?

File: src/maxtext/input_pipeline/multihost_dataloading.py (hunk 3 above)
Comment (SujeethJinesh) [reply src/maxtext/input_pipeline/multihost_dataloading.py:285]: Specialization is used to add more information about inputs and outputs. It has minor performance improvements for the first time the run is done: https://docs.jax.dev/en/latest/notebooks/colocated-python.html#specialization

File: src/maxtext/common/checkpointing.py (hunk 1 above)
Comment (aireenmei) [reply src/maxtext/common/checkpointing.py:122]: So the RemoteIteratorWrapper and RemoteIterator are always used together to support colocated python, independent of elastic.
The current logic is:
```
RemoteIteratorWrapper + RemoteIterator -> colocated python
ElasticIterator -> elastic
```
high level for the code (from line 442 onwards in `grain_data_processing.py`)
```
if colocated:  # Pathways + colocated
    if elastic:
         processing_fn = elastic_processing_fn
    else:
         processing_fn = regular_processing_fn
    return processing_fn + RemoteIteratorWrapper + RemoteIterator
else:  # McJAX, Pathways w/o colocated
    if elastic:
         iterator = ElasticIterator
    else:
         iterator = regular GrainIterator
    return iterator

```

File: src/maxtext/input_pipeline/multihost_dataloading.py (hunk 3 above)
Comment (aireenmei) [reply src/maxtext/input_pipeline/multihost_dataloading.py:285]: Synced offline that specialization is recommended for using colocated python function. In my case, colocated python classes are used.

**Review by SujeethJinesh — APPROVED** [review]
LGTM from Colocated Python Perspective!

**Comment by lukebaumann** [comment]
LGTM from a high level. Did not provide a detailed code review.

**Review by SurbhiJainUSC — APPROVED** [review]
